dp/drivers/i40e.c

```diff
@@ -128,10 +158,186 @@ static int reta_update(struct ix_rte_eth_dev *dev, struct rte_eth_rss_reta *reta
+		I40E_RDT_THRESH) {
+		rxq->tail = rxq->head + rxq->len - 1;
+
+		/* inform HW that more descriptors have become available */
+		I40E_PCI_REG_WRITE(rxq->rdt_reg_addr,
+					(rxq->tail & (rxq->len - 1)));
+	}
+
+	return nb_descs;
+}
+
+/**
+ * rx_queue_setup - prepares an RX queue
+ * @dev: the ethernet device
+ * @queue_idx: the queue number
+ * @numa_node: the desired NUMA affinity, or -1 for no preference
+ * @nb_desc: the number of descriptors to create for the ring
+ *
+ * Returns 0 if successful, otherwise failure.
+ */
 static int rx_queue_setup(struct ix_rte_eth_dev *dev, int queue_idx, int numa_node, uint16_t nb_desc)
 {
-	panic("rx_queue_setup not implemented\n");
+	void *page;
+	machaddr_t page_phys;
+	int ret;
+	struct rx_queue *rxq;
+
+	/*
+	 * The number of receive descriptors must not exceed hardware
+	 * maximum and must be a multiple of I40E_RING_BASE_ALIGN.
+	 */
+	BUILD_ASSERT(align_up(sizeof(struct rx_queue), I40E_RING_BASE_ALIGN) +
+			 align_up(sizeof(union i40e_rx_desc) * I40E_MAX_RING_DESC, I40E_RING_BASE_ALIGN) +
+			 sizeof(struct rx_entry) * I40E_MAX_RING_DESC < PGSIZE_2MB);
+
+	/*
+	 * Additionally, for purely software performance optimization reasons,
+	 * we require the number of descriptors to be a power of 2.
+	 */
+	if (nb_desc & (nb_desc - 1))
+		return -EINVAL;
+
+	/* NOTE: This is a hack, but it's the only way to support late/lazy
+	 * queue setup in DPDK; a feature that IX depends on. */
+	rte_eth_devices[dev->port].data->nb_rx_queues = queue_idx + 1;
+
+	ret = rte_eth_rx_queue_setup(dev->port, queue_idx, nb_desc, numa_node, &rx_conf, dpdk_pool);
+	if (ret < 0)
+		return ret;
+	if (numa_node == -1) {
+		page = mem_alloc_page_local(PGSIZE_2MB);
+		if (page == MAP_FAILED)
+			return -ENOMEM;
+	} else {
+		page = mem_alloc_page(PGSIZE_2MB, numa_node, MPOL_BIND);
+		if (page == MAP_FAILED)
+			return -ENOMEM;
+	}
+	memset(page, 0, PGSIZE_2MB);
+
+	/* hijack dpdk usual memory with our own from bigpages:*/
+	rxq = (struct rx_queue *) page;
+
+	rxq->ring  = (void *)((uintptr_t) page +
+		align_up(sizeof(struct rx_queue), I40E_RING_BASE_ALIGN));
+	rxq->ring_entries = (struct rx_entry *)((uintptr_t) rxq->ring +
+		align_up(sizeof(union i40e_rx_desc) * nb_desc, I40E_RING_BASE_ALIGN));
+
+	rxq->len = nb_desc;
+	rxq->head = 0;
+	rxq->tail = rxq->len - 1;
+
+	ret = mem_lookup_page_machine_addr(page, PGSIZE_2MB, &page_phys);
+	if (ret)
+		goto err;
+
+	rxq->ring_physaddr = page_phys +
+				 align_up(sizeof(struct rx_queue), I40E_RING_BASE_ALIGN);
+	log_err("queue_setup : phys %p\n", rxq->ring_physaddr);
+	/* TODO: this is now done in dev_start (need access to dpdk's queue) */
+	//rxq->reg_idx = drxq->reg_idx;
+	//rxq->rdt_reg_addr = (uint32_t*) drxq->qrx_tail;
+
+	rxq->erxq.poll = i40e_rx_poll;
+	dev->data->rx_queues[queue_idx] = &rxq->erxq;
+	rxq->erxq.drv_rx_queue = rxq;
+	/* release the dpdk memory location and all its buffers*/
+	//i40e_dev_rx_queue_release(drxq);
```

Comment:
That's a good idea. Unless it breaks something in a weird way. If it works correctly, then you could to do the same in the ixgbe driver later on.